exp/runtime/api/v1beta1/extensionconfig_types.go

```diff
@@ -0,0 +1,195 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ANCHOR: ExtensionConfigSpec
+
+// ExtensionConfigSpec defines the desired state of ExtensionConfig.
+type ExtensionConfigSpec struct {
+	// ClientConfig defines how to communicate with ExtensionHandlers.
+	ClientConfig ClientConfig `json:"clientConfig"`
+
+	// NamespaceSelector decides whether to run the webhook on an object based
+	// on whether the namespace for that object matches the selector.
+	// Default to the empty LabelSelector, which matches everything.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+// ClientConfig contains the information to make a client
+// connection with an ExtensionHandler.
+type ClientConfig struct {
+	// URL gives the location of the ExtensionHandler, in standard URL form
+	// (`scheme://host:port/path`). Exactly one of `url` or `service`
+	// must be specified.
+	//
+	// The `host` should not refer to a service running in the cluster; use
+	// the `service` field instead.
+	//
+	// The scheme should be "https"; the URL should begin with "https://".
+	// "http" is supported for insecure development purposes only.
+	//
+	// A path is optional, and if present may be any string permissible in
+	// a URL. If a path is set it will be used as prefix and the hook-specific
+	// path will be appended.
+	//
+	// Attempting to use a user or basic auth e.g. "user:password@" is not
+	// allowed. Fragments ("#...") and query parameters ("?...") are not
+	// allowed either.
+	//
+	// +optional
+	URL *string `json:"url,omitempty"`
+
+	// Service is a reference to the Kubernetes service for the ExtensionHandler.
+	// Either `service` or `url` must be specified.
+	//
+	// If the ExtensionHandler is running within a cluster, then you should use `service`.
+	//
+	// +optional
+	Service *ServiceReference `json:"service,omitempty"`
+
+	// CABundle is a PEM encoded CA bundle which will be used to validate the ExtensionHandler's server certificate.
+	// +optional
+	CABundle []byte `json:"caBundle,omitempty"`
+}
+
+// ServiceReference holds a reference to a Kubernetes Service.
+type ServiceReference struct {
+	// Namespace is the namespace of the service.
+	Namespace string `json:"namespace"`
+
+	// Name is the name of the service.
+	Name string `json:"name"`
+
+	// Path is an optional URL path which will be sent in any request to
+	// this service. If a path is set it will be used as prefix and the hook-specific
+	// path will be appended.
+	// +optional
+	Path *string `json:"path,omitempty"`
+
+	// Port is the port on the service that hosting ExtensionHandler.
+	// Default to 8443.
+	// `port` should be a valid port number (1-65535, inclusive).
```

Comment:
Any reason not to use kubebuilder validations for this? FWIW having it in the godoc is also useful for `kubectl explain`, but having in openapi can also help 
```suggestion
	// Default to 8443.
	// `port` should be a valid port number (1-65535, inclusive).
	// +kubebuilder:default:=8443
	// +kubebuilder:validation:Maximum:=65535
	// +kubebuilder:validation:Minimum:=1
```
I would also question whether it needs to be a pointer or not

Reply:
Doesn't have to be a pointer if we want to treat 0 as unset which would be possible.

Our general policy regarding kubebuilder defaulting/validation was that we favored implementing it in a webhook so we can unit test it accordingly. I think we can't do that with OpenAPI schema. 

But we should definitely discuss this.

Reply:
OpenAPI is limited in what it can do for sure, but you can write unit tests for it with envtest at least, which of course has an overhead which we will want to discuss, is that something we are happy to accept

Reply:
Yup sounds like that's the tradeoff. I think the current situation is that the way we setup envtest if we add those unit tests to the api package we end up with a cyclic dependency. This can be resolved by placing the env test unit tests in a separate package (we did it here because those were really important: https://github.com/kubernetes-sigs/cluster-api/tree/main/internal/webhooks/test)

Reply:
Immutability is one important lack in current CRD OpenAPI validation.

Reply:
Also the fact that any conceptually required field that you want to default needs to be declared optional for client side validation to not complain e.g kubectl, is suboptimal.